### PR TITLE
update vcfanno

### PR DESCRIPTION
**Pull request by matthdsm**
Update vcfanno module

- now takes explicit `lua` file input
- now takes a list of reference file instead of a directory
- stage input vcfs under input to avoid name collisions

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by pinin4fjords** [comment]
> Any idea why this module would exit with code `0` ? I'm getting output and everything seems to finish like it should 🤷🏻

If it helps, the error from the logs is:

```
N E X T F L O W  ~  version 22.10.3
Launching `./tests/modules/nf-core/vcfanno/main.nf` [sleepy_thompson] DSL2 - revision: c6778f9d67
[3f/a3a712] Submitted process > test_vcfanno:VCFANNO (test_compressed)
Error executing process > 'test_vcfanno:VCFANNO (test_compressed)'

Caused by:
  Missing output file(s) `*_annotated.vcf` expected by process `test_vcfanno:VCFANNO (test_compressed)`

Command executed:

  vcfanno \
      -p 2 \
       \
       \
      vcfanno.toml \
      input/test.vcf.gz \
      > test_compressed.vcf
  
  cat <<-END_VERSIONS > versions.yml
  "test_vcfanno:VCFANNO":
      vcfanno: $(echo $(vcfanno 2>&1 | grep version | cut -f3 -d' ' ))
  END_VERSIONS

Command exit status:
  0

Command output:
  (empty)
...
```

**Comment by matthdsm** [comment]
I got my error during `create-test-yml` so that shouldn't matter 🤔

File: modules/nf-core/vcfanno/meta.yml [hunk 1]
```diff
@@ -32,14 +32,15 @@ input:
...
+  - lua:
       type: file
-      description: |
-        This directory contains referenced files in the TOML config,
-        and the corresponding indicies e.g. exac.vcf.gz + exac.vcf.gz.tbi,
-        with exception to the lua file.
+      description: Lua file for custom annotations
+      patter: "*.lua"
+  - resources:
+      type: file
```
Comment (pinin4fjords) [inline modules/nf-core/vcfanno/meta.yml:42]:
```suggestion
      type: list
```

File: modules/nf-core/vcfanno/meta.yml [hunk 2]
```diff
@@ -32,14 +32,15 @@ input:
...
       pattern: "*.toml"
-  - resource_dir:
+  - lua:
       type: file
-      description: |
-        This directory contains referenced files in the TOML config,
-        and the corresponding indicies e.g. exac.vcf.gz + exac.vcf.gz.tbi,
-        with exception to the lua file.
+      description: Lua file for custom annotations
+      patter: "*.lua"
```
Comment (pinin4fjords) [inline modules/nf-core/vcfanno/meta.yml:40]:
```suggestion
      pattern: "*.lua"
```

**Comment by matthdsm** [comment]
Damn @pinin4fjords , that was really a rookie mistake 😮 and I kept missing it too 😭 
Anyway, fixed now. Thanks for the helping eye!

File: tests/modules/nf-core/vcfanno/test.yml [hunk 3]
```diff
@@ -1,23 +1,15 @@
 - name: vcfanno test_vcfanno
-  command: nextflow run ./tests/modules/nf-core/vcfanno -entry test_vcfanno -c ./tests/config/nextflow.config  -c ./tests/modules/nf-core/vcfanno/nextflow.config
+  command: nextflow run ./tests/modules/nf-core/vcfanno -entry test_vcfanno -c ./tests/config/nextflow.config -c ./tests/modules/nf-core/vcfanno/nextflow.config
   tags:
     - vcfanno
   files:
-    - path: output/untar/vcfanno_grch38_module_test/exac.vcf.gz
-      md5sum: eba4b55d81148d9c316cda68a5c0ad6d
-    - path: output/untar/vcfanno_grch38_module_test/exac.vcf.gz.tbi
```
Comment (pinin4fjords) [inline tests/modules/nf-core/vcfanno/test.yml:8]: Was the .tbi output included by mistake previously?

File: modules/nf-core/vcfanno/main.nf [hunk 4]
```diff
@@ -22,15 +23,15 @@ process VCFANNO {
...
     """
-    ln -sf $resource_dir/* \$(pwd)
-
     vcfanno \\
         -p $task.cpus \\
         $args \\
+        $lua \\
         $toml \\
         $vcf \\
-        > ${prefix}_annotated.vcf
```
[the remark this answers is not in the document]
Comment (pinin4fjords) [reply modules/nf-core/vcfanno/main.nf:33]: Don't forget to apply the same fix to the stub!

File: tests/modules/nf-core/vcfanno/test.yml (hunk 3 above)
Comment (matthdsm) [reply tests/modules/nf-core/vcfanno/test.yml:8]: Neither of those files should've been included, as they're inputs

File: modules/nf-core/vcfanno/main.nf [hunk 5]
```diff
@@ -8,12 +8,13 @@ process VCFANNO {
         'quay.io/biocontainers/vcfanno:0.3.3--h9ee0642_0' }"
 
     input:
-    tuple val(meta), path(vcf), path(tbi)
+    tuple val(meta), path(vcf, stageAs: "input/*"), path(tbi, stageAs: "input/*")
```
Comment (FriederikeHanssen) [inline modules/nf-core/vcfanno/main.nf:11]: why is the stageAs needed here?

**Comment by FriederikeHanssen** [comment]
> Because a vcf input with the same $prefix can cause a name collision Op 28 nov. 2022 om 16:43 heeft FriederikeHanssen ***@***.***> het volgende geschreven: ﻿ @FriederikeHanssen commented on this pull request.
> ________________________________ In modules/nf-core/vcfanno/main.nf<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnf-core%2Fmodules%2Fpull%2F2560%23discussion_r1033706073&data=05%7C01%7CMatthias.DeSmet%40ugent.be%7C0f5f994f87c14a5bd39f08dad15742f8%7Cd7811cdeecef496c8f91a1786241b99c%7C1%7C0%7C638052469967584710%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=PV%2Ba98%2Bp096i712GeTNmQEVf%2BTlyDWGGG49l7Ye8yJ4%3D&reserved=0>:
> @@ -8,12 +8,13 @@ process VCFANNO {
> 'quay.io/biocontainers/vcfanno:0.3.3--h9ee0642_0' }" input: - tuple val(meta), path(vcf), path(tbi) + tuple val(meta), path(vcf, stageAs: "input/*"), path(tbi, stageAs: "input/*") why is the stageAs needed here? — Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnf-core%2Fmodules%2Fpull%2F2560%23pullrequestreview-1196047520&data=05%7C01%7CMatthias.DeSmet%40ugent.be%7C0f5f994f87c14a5bd39f08dad15742f8%7Cd7811cdeecef496c8f91a1786241b99c%7C1%7C0%7C638052469967584710%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=z1lQtKac87OQ5IwEpqP5mWl5LVb0RoLkyCJZJAmsAEg%3D&reserved=0>, or unsubscribe<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAC2NHEGR7N76XVBVCOV4CGDWKTHJBANCNFSM6AAAAAASNJWY2Y&data=05%7C01%7CMatthias.DeSmet%40ugent.be%7C0f5f994f87c14a5bd39f08dad15742f8%7Cd7811cdeecef496c8f91a1786241b99c%7C1%7C0%7C638052469967740887%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pCEdCHC64IBBrrJWv3MUOrLcJcTL3NxQBUJ7xxvtwtk%3D&reserved=0>. You are receiving this because you were assigned.Message ID: ***@***.***>

you are only processing one vcf at a time in this module at least according to the meta, right? NExtflow doesn't overwrite in and output files: https://www.nextflow.io/docs/latest/process.html#dynamic-output-file-names

**Comment by matthdsm** [comment]
Great, missed that somehow. I'll remove the stageAs then

File: modules/nf-core/vcfanno/main.nf (hunk 5 above)
Comment (matthdsm) [reply modules/nf-core/vcfanno/main.nf:11]:
```suggestion
    tuple val(meta), path(vcf), path(tbi)
```
Review comments

**Comment by nvnieuwk** [comment]
A piece of code to check if the input file has the same name as the output file could also be useful :)

**Comment by matthdsm** [comment]
Thats kind of moot when nextflow checks that for itself 🤷🏻

**Comment by nvnieuwk** [comment]
Okay, yeah fair point!

**Comment by nvnieuwk** [comment]
But it only checks after the module has been run, which could save some time?

**Pull request merged**